src/config/compiler.rs

```diff
@@ -1,7 +1,48 @@
 use super::{builder::ConfigBuilder, validation, Config, TransformOuter};
 use indexmap::IndexMap;
 
-pub fn compile(mut builder: ConfigBuilder) -> Result<(Config, Vec<String>), Vec<String>> {
+/// Compile defers to either building a "stub" config, deferring the rest to a provider...
+/// or configuring components directly.
+#[cfg(feature = "providers")]
+pub fn compile(builder: ConfigBuilder) -> Result<(Config, Vec<String>), Vec<String>> {
+    if builder.provider.is_some() {
+        compile_provider(builder)
+    } else {
+        compile_components(builder)
+    }
+}
```

Comment:
I understand, although a key distinction with a provider is that a config isn't in a "ready' state until we've received the first configuration back. This branch is intended to accommodate that scenario.

We also want to check that a `[provider]` and other topology aren't _both_ present in the config, which is another reason for it.

This could totally be moved to a separate concern, but it's not immediately clear to me how we'd want to do that, if we also need to validate top-level keys such as API, data_dir, etc.

Open to thoughts here -- like most design decisions in this PR, this can probably go one of several ways, and I'm not sure which is preferred!